lightning/src/ln/channelmanager.rs

```diff
@@ -3478,6 +3505,74 @@ impl<ChanSigner: ChannelKeys, M: Deref + Sync + Send, T: Deref + Sync + Send, K:
 	}
 }
 
+/// Used to signal to the ChannelManager persister that the manager needs to be re-persisted to
+/// disk/backups, through `wait_timeout`.
+pub struct PersistenceNotifier {
+	/// Used to signal to the ChannelManager persister that the manager needs to be re-persisted to
+	/// disk/backups. Users won't access the persistence_lock directly, but rather wait on its bool
+	/// using wait_timeout.
+	#[cfg(any(test, feature = "_test_utils"))]
+	pub persistence_lock: Arc<(Mutex<bool>, Condvar)>,
+	#[cfg(not(any(test, feature = "_test_utils")))]
+	persistence_lock: Arc<(Mutex<bool>, Condvar)>,
+}
+
+impl PersistenceNotifier {
+	fn new() -> Self {
+		Self {
+			persistence_lock: Arc::new((Mutex::new(false), Condvar::new())),
+		}
+	}
+
+	/// Blocks until ChannelManager needs to be persisted or a timeout is reached. If no timeout is
+	/// provided, this function will block indefinitely until persistence is needed. It returns a bool
+	/// indicating whether persistence is necessary (which will always be true if max_wait is None).
+	#[cfg(feature = "allow_wallclock_use")]
+	pub fn wait_timeout(&self, max_wait: Option<Duration>) -> bool {
+		let current_time = Instant::now();
+		let mut time_to_wait = max_wait;
+		loop {
+			let mutcond = Arc::clone(&self.persistence_lock);
+			let &(ref mtx, ref cvar) = &*mutcond;
+			let mut guard = mtx.lock().unwrap();
+
+			let result = match time_to_wait {
+				Some(wait) => cvar.wait_timeout(guard, wait).unwrap().0,
+				None => cvar.wait(guard).unwrap(),
+			};
+			// Due to spurious wakeups that can happen on `wait_timeout`, here we need to check if the
+			// desired wait time has actually passed, and if not then restart the loop with a reduced wait
+			// time. Note that this logic can be highly simplified through the use of
+			// `Condvar::wait_while` and `Condvar::wait_timeout_while`, if and when our MSRV is raised to
+			// 1.42.0.
+			let elapsed = current_time.elapsed();
+			guard = result;
+			let saved_res = *guard;
+			*guard = false;
+			if saved_res || (max_wait.is_some() && elapsed >= max_wait.unwrap()) {
+				return saved_res;
+			} else if max_wait.is_some() {
+				let remaining_time = max_wait.unwrap().checked_sub(elapsed);
+				if remaining_time.is_some() {
+					time_to_wait = Some(remaining_time.unwrap());
+				} else {
+					// There was actually no time remaining, so just return the result.
+					return saved_res;
+				}
+			}
+		}
```

Comment:
Instead of using `remaining_time` as an intermediary variable, you can directly assign to `time_to_wait`.

I'd also consider using `match` to separate each case. YMMV, but I found it more readable since it declutters the `unwrap()`s and separates out each case. It also allows you to reset the bool only when needed.

```rust
if result {
	*guard = false;
	return result;
}

match max_wait {
	None => continue,
	Some(max_wait) if elapsed >= max_wait => return result,
	Some(max_wait) => {
		time_to_wait = max_wait.checked_sub(elapsed);
		match time_to_wait {
			None => return result,
			Some(_) => continue,
		}
	},
}
```